### PR TITLE
Mirror unloading of units to starting

### PR DESCRIPTION
**Pull request by miekg**
First trigger stop then unload the unit in a separate loop. This mimics
how we start the units in the first place.

Also unify the logging for the 2 operations.

Signed-off-by: Miek Gieben <miek@miek.nl>

### CODE REVIEW

File: systemd/pod.go [hunk 1]
```diff
@@ -289,13 +289,18 @@ func (p *P) UpdatePod(ctx context.Context, pod *corev1.Pod) error {
...
 		}
+		unitsToUnload = append(unitsToUnload, name)
+	}
+
+	for _, name := range unitsToUnload {
 		if err := p.m.Unload(name); err != nil {
-			klog.Warningf("Failed to unload: %s", err)
+			klog.Warningf("Failed to unload unit %q: %s", name, err)
 		}
 		klog.Infof("Deleted unit [%q] successfully", name)
```
Comment (pires) [inline systemd/pod.go:305]: If the unit can't be unloaded did we really delete it?

**Review by pires — COMMENTED** [review]
I'm arguing for the current `DeletePod` behavior to stay as is and for `CreatePod` to change instead as I don't see the need for adding yet another iteration, and another array allocation and dynamic growth, for an array that is already being iterated upon.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [virtual-kubelet/systemk] Mirro..." ]
>@pires commented on this pull request.
>
>I'm arguing for the current DeletePod behavior to stay as is and for CreatePod
>to change instead as I don't see the need for adding yet another iteration, and
>another array allocation and dynamic growth, for an array that is already being
>iterated upon.

that won't work, as Before unit (init containers) don't get scheduled correctly.

**Comment by pires** [comment]
Can you develop further what problem you are observing?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [virtual-kubelet/systemk] Mirro..." ]
>Can you develop further what problem you are observing?

init container are references by the main ones with an 'After:', it the unit isn't loaded
stuff didn't work - so this when developing this feature - loading and *then* TriggerStart
solved this problem

**Comment by pires** [comment]
Are you saying the `Load` is asynchronous and not really assuring that a unit was indeed loaded?

This doesn't seem to be deterministic behavior and _we_ are just lucky that in between iterating the two lists there's enough time for the dependency units to actually be loaded. And I'm sorry to say, I don't like it. Worst-case scenario, we should verify if all units are loaded before triggering them.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [virtual-kubelet/systemk] Mirro..." ]
>Are you saying the Load is asynchronous and not really assuring that a unit was
>indeed loaded?

Well, if you have before/after and thus depend on other units, it can't synchronous?

>This doesn't seem to be deterministic behavior and we are just lucky that in
>between iterating the two lists there's enough time for the dependency units to
>actually be loaded. And I'm sorry to say, I don't like it. Worst-case scenario,
>we should verify if all units are loaded before triggering them.

We also don't have enough evidence for this. All I can say is that I hit a snag, did this
and now it works (with the little testing that went into it)

**Comment by pires** [comment]
For instance, iterating the two arrays is not guaranteed to be ordered so what happens when one triggers a unit that depends on another one that hasn't been triggered?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [virtual-kubelet/systemk] Mirro..." ]
>For instance, iterating the two arrays is not guaranteed to be ordered so what
>happens when one triggers a unit that depends on another one that hasn't been
>triggered?

They _all_ need to be loaded, actually triggering doesn't matter here, because the
'After:' dependency has been seen by systemd. So that unit referenced in 'After' will be
done first.

You just can't load and trigger at the same time (or at least in the same loop), because
that dep can't be established.



/Miek

--
Miek Gieben

**Comment by pires** [comment]
Ah, that's so obvious it hurts. Sorry!

**Review by pires — APPROVED** [review]

**Pull request merged**